README
				
                                  Gdadin
		     http://freaknet.org/alpt/src/gdadin
               GUI to Draw Algorithms Designed with Inkscape


Gdadin is an Inkscape effect. It creates complex drawings from simple rules
The rules are defined with an intuitive GUI based on the Copy&Paste principle.

The Gdadin concept and its main design have been heavily inspired by
Cfdg ( http://www.contextfreeart.org/ ), a programming language used to
compose and create drawings. Almost all drawings created with Cfdg can be
reproduced with Gdadin.

--
  0. Examples and Lessons
  1. General workflow
  2. Recursion and cycles
	2.1  Size rule 
  3. Shape call parameters
	3.1  Recursion depth
	3.2  Colors
  4. Random shapes
	4.1  Random seed
  5. Limits
  6. Tips
--

0.   Examples and Lessons

For examples look in examples/ or here: http://freaknet.org/alpt/src/gdadin/examples/

You can find video lessons here:  http://freaknet.org/alpt/src/gdadin/lessons/

1.   General workflow

A Gdadin shape is a group of objects created with inkscape, f.e. a square, a
star, text. 
To define a new shape:
	- select the objects
	- Use "Effects->Gdadin->Define" a new shape
	- The defined shape will be exactly on top of the selected objects.
	  The shape contains a copy of the objects therefore you can freely
	  delete the originals.

Defined shapes can be copied and pasted like any other normal object. 
In order to edit a defined shape, you must enter inside its group: 
"Right click->Enter" group (and remember to exit: "Right click -> Go to Parent").

You can paste a shape A inside another shape B:
	- select the shape A
	 (be sure to select its whole group and not just some single elements
	  of the shape)
	- copy it
	- enter in the shape B (enter group)
	- paste
	- exit from the shape B

NOTE: The pasted shape is just a reference to the original shape: the rendering 
will delete the reference and substitute it with the original. Therefore,
	1) if you want to modify the pasted shape you have to modify just the
	   original one.
	2) you can't delete the shape if it has been referenced by another
	   shape and you want to render it.

To define a recursive shape: copy and paste the shape inside itself. In other
words,
	- copy the shape A
	- enter inside the shape A
	- paste it
	- exit from the shape A

Finally, select the final shapes you'd like to render and use 
"Effects->Gdadin->Render selected shapes".

Here below we'll explain how to use cycles, color variations and random
shapes.


2.   Recursion and cycles

2.1  Size rule 

Suppose you've defined a recursive shape A and you render it.
If it definitively grows smaller, Gdadin will halt the recursion when the
shape becomes too small to be seen.
Example:
	- draw a rectangle, select it and define a new shape.
	- copy the shape, enter inside the shape, paste it
	- shrink it 
	- go to parent, select the shape, render it

When it isn't possible to apply the size rule you have to manually set a
limit. The render dialog offers you the "Maximum number of shapes" option.
It is the absolute maximum number of shape substitution that will be performed
during the rendering.
This option is useful to experiment with complex drawings: with a low number
you can render the drawing quickly, hit CTRL+z, and return to edit.

3.   Shape call parameters

3.1  Recursion depth

The recursion depth parameter can be assigned to a referenced/pasted shape,
generally a recursive one. It specifies how many child shapes of the referenced
shape will be created. This is useful to create cycles.
For example:
	- Let's create a tower using the size rule:
		- draw a rectangle, select it
		- define new shape, copy the shape, enter inside the shape, paste
		- shrink the pasted shape and move it on top of the big
		  rectangle
		- go to parent
	- Now we'll rotate the tower three times:
		- select the tower shape, copy it
		- enter in the tower shape, paste in place (CTRL+ALT+v)
		- rotate it
		- keep it selected, and open the dialog
		  "Effects->Gdadin->Set the shape call parameter"
		- Set "Recursion depth" to 2.
		- go to parent
	- Let's render it (without any limit, i.e. "maximum number of shape"
	  set to 0) 
	- You should get the tower rotated three times.

You've just seen how to create a cycle.

3.2  Colors

These parameters define how the color of a referenced shape will change over
time, i.e. the new color that its same child shape will have.

The parameters are Hue, Saturation, Brightness, Opacity (If you don't know
what they are, just draw an object, hit CTRL+SHIFT+f "Fill and Stroke", go to the 
HSL tab, and experiment)

We'll describe just the Hue parameter:

	- The "Relative hue change" is an integer in the range [-1000, 1000].
	  It is used to calculate the new Hue of the child of the referenced
	  shape:
	  	 newhue = referenced_shape_hue + relative_hue_change/1000.0

	  Note: relative_hue_change can be also negative.
	  Note: the Hue is a real parameter in the range [0, 1]. Thus,
	  	relative_hue_change = 1 is a very little increment
		relative_hue_change = 10 is a smooth, gradual increment
		relative_hue_change = 100 is a steep increment
		
	- "Maximum hue value" sets a ceil for the newhue:
		
		if newhue >= maximum_hue_value/1000.0 then
			newhue = maximum_hue_value/1000.0

	  In other words, when the hue reaches the maximum_hue_value, the
	  variation stops: the new childs will have the same hue.

	  You can disable it by choosing a negative value, f.e. -1.
	  This is the default.
	  In this case, the hue will change cyclically, f.e. 

	  	initial_hue = 312,  relative_hue_change = 10
		312 322 332  ... 982 992  2  12  22  ...

	  	initial_hue = 312,  relative_hue_change = -10
		312 302 292  ... 32 22 12 2 992 982 ...

	- "Minimum hue value". The same as before, but sets a floor instead.

4.   Random shapes

When you define a new shape, you can set two (optional) parameters: 
	1) The shape name
	2) The shape weight
Suppose now that you define more than one shape with the same name and you
reference, i.e. copy and paste, at least one of them.
In this case, during the rendering, Gdadin will resolve the reference by
choosing randomly one of the shapes with the same name.

The weight specifies how much probability has each the shape to be chosen.

Hint: random shapes are useful to give an organic feel to the drawing.

4.1  Random seed

Naturally, each time you render a random shape, you'll get a different
results. 
In order to fix a unique result, you can set an arbitrary alphanumeric string
in the "Random seed" parameter of the render dialog.

You can check what seed has been used during the rendering by looking in the
XML editor:
	- select the final rendered shape
	- hit CTRL+SHIFT+x
	- read the "inkscape:seed" attribute.
This procedure is useful if you want to reproduce a same random rendering.

5.   Limits	

It's slow: very complicated shapes, with more than thousand shapes can take a
while to render and can easily fill up memory.

6.   Tips

	1) use a low Max Number of Shapes to test the drawing

	2) save before rendering

	3) After a long rendering: 
		- don't touch anything
		- export to bitmap (SHIFT+CTRL+e) as soon as possible.
		  As export area choose "Page", because after a rendering
		  Gdadin will automatically resize the page to contain the
		  final drawing.
	 	- you can export to bitmap by commandline (it's better):
			- save the rendering to file.svg
			- inkscape file.svg -C -e out.png -w 640 -h 480

	4) Learn to use the XML editor

7.   Authors

Andrea Lo Pumo  aka  AlpT (@freaknet.org)



ChangeLog
- 0.0.1

* Initial release: it works.


stylevariation.py
#!/usr/bin/env python 
#Note: taken from inkscape/extensions/coloreffect,py (with minor changes)
'''
Copyright (C) 2006 Jos Hirth, kaioa.com
Copyright (C) 2007 Aaron C. Spike

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
'''

import simplestyle
import pdb

color_props_fill=('fill:','stop-color:','flood-color:','lighting-color:')
color_props_stroke=('stroke:',)
color_props = color_props_fill + color_props_stroke
opacity_props = 'opacity:'

def apply_colmod(self, nodeList, hsbcoeffs, opacoeff, hsbo_cf):
  for n in nodeList:
      changeStyle(self, n, hsbcoeffs, opacoeff, hsbo_cf)
      apply_colmod(self, n, hsbcoeffs, opacoeff, hsbo_cf)

def changeStyle(self, node, hsbcoeffs, opacoeff, hsbo_cf):
  if node.attrib.has_key('style'): 
    style=node.get('style') # FIXME: this will break for presentation attributes!
    if style!='':
      #inkex.debug('old style:'+style)
      styles=style.split(';')
      for i in range(len(styles)):
        if styles[i].startswith(opacity_props):
                styles[i]=opacity_props+str(opamod(float(styles[i][len(opacity_props):]), opacoeff, hsbo_cf))
        else:
                for c in range(len(color_props)):
                  if styles[i].startswith(color_props[c]):
                    styles[i]=color_props[c]+process_prop(self, styles[i][len(color_props[c]):], hsbcoeffs, hsbo_cf)

      #inkex.debug('new style:'+';'.join(styles))
      node.set('style',';'.join(styles))

def process_prop(self, col, hsbcoeffs, hsbo_cf):
  #debug('got:'+col)
  if simplestyle.isColor(col):
    c=simplestyle.parseColor(col)
    col='#'+colmod(c[0],c[1],c[2], hsbcoeffs, hsbo_cf)
    #debug('made:'+col)
  if col.startswith('url(#'):
    id = col[len('url(#'):col.find(')')]
    newid = self.newid(self.svg_prefixfromtag(id))
    #inkex.debug('ID:' + id )
    path = '//*[@id="%s"]' % id
    for node in self.document.xpath(path, namespaces=inkex.NSS):
      process_gradient(self, node, newid, hsbcoeffs)
    col = 'url(#%s)' % newid
  return col

def process_gradient(self, node, newid, hsbcoeffs, hsbo_cf):
  newnode = copy.deepcopy(node)
  newnode.set('id', newid)
  node.getparent().append(newnode)
  changeStyle(self, newnode, hsbcoeffs, opacoeff, hsbo_cf)
  for child in newnode:
    changeStyle(self, child, hsbcoeffs, opacoeff, hsbo_cf)
  xlink = inkex.addNS('href','xlink')
  if newnode.attrib.has_key(xlink):
    href=newnode.get(xlink)
    if href.startswith('#'):
      id = href[len('#'):len(href)]
      #inkex.debug('ID:' + id )
      newhref = self.newid(self.svg_prefixfromtag(id))
      newnode.set(xlink, '#%s' % newhref)
      path = '//*[@id="%s"]' % id
      for node in self.document.xpath(path, namespaces=inkex.NSS):
        process_gradient(node, newhref)

hsb_chars = {0: 'h', 1: 's', 2: 'b'}
def colmod(r,g,b, (hue_coeff, sat_coeff, bri_coeff), hsbo_cf):
        hsl = rgb_to_hsl(r/255.0, g/255.0, b/255.0)
        hsl[0] = hsl[0]+hue_coeff
        hsl[1] = hsl[1]+sat_coeff
        hsl[2] = hsl[2]+bri_coeff
        for i in range(len(hsl)):
                fidx = hsb_chars[i]+'f'
                cidx = hsb_chars[i]+'c'
                if hsbo_cf[fidx] >= 0 and hsl[i] <= hsbo_cf[fidx]:
                        hsl[i] = hsbo_cf[fidx]
                if hsbo_cf[cidx] >= 0 and hsl[i] >= hsbo_cf[cidx]:
                        hsl[i] = hsbo_cf[cidx]
                if abs(hsl[i]) > 1.0:
                        hsl[i] -= int(hsl[i])
                if hsl[i] < 0:
                        hsl[i]=1.0+hsl[i]
        rgb = hsl_to_rgb(hsl[0], hsl[1], hsl[2])
        return '%02x%02x%02x' % (rgb[0]*255, rgb[1]*255, rgb[2]*255)

def opamod(opa, opacoeff, hsbo_cf):
        opa+=opacoeff
        if hsbo_cf['of'] >=0 and opa <= hsbo_cf['of']:
                opa = hsbo_cf['of']
        if hsbo_cf['oc'] >=0 and opa >= hsbo_cf['oc']:
                opa = hsbo_cf['oc']
        if abs(opa) > 1.0:
                opa -= int(opa)
        if opa < 0:
                opa=1.0+opa
        return opa

def rgb_to_hsl(r, g, b):
  rgb_max = max (max (r, g), b)
  rgb_min = min (min (r, g), b)
  delta = rgb_max - rgb_min
  hsl = [0.0, 0.0, 0.0]
  hsl[2] = (rgb_max + rgb_min)/2.0
  if delta == 0:
      hsl[0] = 0.0
      hsl[1] = 0.0
  else:
      if hsl[2] <= 0.5:
          hsl[1] = delta / (rgb_max + rgb_min)
      else:
          hsl[1] = delta / (2 - rgb_max - rgb_min)
      if r == rgb_max:
          hsl[0] = (g - b) / delta
      else:
          if g == rgb_max:
              hsl[0] = 2.0 + (b - r) / delta
          else:
              if b == rgb_max:
                  hsl[0] = 4.0 + (r - g) / delta
      hsl[0] = hsl[0] / 6.0
      if hsl[0] < 0:
          hsl[0] = hsl[0] + 1
      if hsl[0] > 1:
          hsl[0] = hsl[0] - 1
  return hsl

def hue_2_rgb (v1, v2, h):
  if h < 0:
      h += 6.0
  if h > 6:
      h -= 6.0
  if h < 1:
      return v1 + (v2 - v1) * h
  if h < 3:
      return v2
  if h < 4:
      return v1 + (v2 - v1) * (4 - h)
  return v1

def hsl_to_rgb (h, s, l):
  rgb = [0, 0, 0]
  if s == 0:
      rgb[0] = l
      rgb[1] = l
      rgb[2] = l
  else:
      if l < 0.5:
          v2 = l * (1 + s)
      else:
          v2 = l + s - l*s
      v1 = 2*l - v2
      rgb[0] = hue_2_rgb (v1, v2, h*6 + 2.0)
      rgb[1] = hue_2_rgb (v1, v2, h*6)
      rgb[2] = hue_2_rgb (v1, v2, h*6 - 2.0)
  return rgb


T.py
#!/usr/bin/env python
#Note: taken from inkscape/extensions/simpletransform.py (with minor changes)
'''
Copyright (C) 2006 Jean-Francois Barraud, barraud@math.univ-lille1.fr

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
barraud@math.univ-lille1.fr

This code defines several functions to make handling of transform
attribute easier.
'''
import inkex, cubicsuperpath 
import math, re

def parseTransform(transf,mat=[[1.0,0.0,0.0],[0.0,1.0,0.0]]):
    if transf=="" or transf==None:
        return(mat)
    result=re.match("(translate|scale|rotate|skewX|skewY|matrix)\(([^)]*)\)",transf)
#-- translate --
    if result.group(1)=="translate":
        args=result.group(2).split(",")
        dx=float(args[0])
        if len(args)==1:
            dy=0.0
        else:
            dy=float(args[1])
        matrix=[[1,0,dx],[0,1,dy]]
#-- scale --
    if result.groups(1)=="scale":
        args=result.group(2).split(",")
        sx=float(args[0])
        if len(args)==1:
            sy=sx
        else:
            sy=float(args[1])
        matrix=[[sx,0,0],[0,sy,0]]
#-- rotate --
    if result.groups(1)=="rotate":
        args=result.group(2).split(",")
        a=float(args[0])*math.pi/180
        if len(args)==1:
            cx,cy=(0.0,0.0)
        else:
            cx,cy=args[1:]
        matrix=[[math.cos(a),-math.sin(a),cx],[math.sin(a),math.cos(a),cy]]
#-- skewX --
    if result.groups(1)=="skewX":
        a=float(result.group(2))*math.pi/180
        matrix=[[1,math.tan(a),0],[0,1,0]]
#-- skewX --
    if result.groups(1)=="skewX":
        a=float(result.group(2))*math.pi/180
        matrix=[[1,0,0],[math.tan(a),1,0]]
#-- matrix --
    if result.group(1)=="matrix":
        a11,a21,a12,a22,v1,v2=result.group(2).split(",")
        matrix=[[float(a11),float(a12),float(v1)],[float(a21),float(a22),float(v2)]]
    
    matrix=composeTransform(mat,matrix)
    if result.end()<len(transf):
        return(parseTransform(transf[result.end():],matrix))
    else:
        return matrix

def formatTransform(mat):
    return("matrix(%f,%f,%f,%f,%f,%f)"%(mat[0][0],mat[1][0],mat[0][1],mat[1][1],mat[0][2],mat[1][2]))

def composeTransform(M1,M2):
    a11=M1[0][0]*M2[0][0]+M1[0][1]*M2[1][0]
    a12=M1[0][0]*M2[0][1]+M1[0][1]*M2[1][1]
    a21=M1[1][0]*M2[0][0]+M1[1][1]*M2[1][0]
    a22=M1[1][0]*M2[0][1]+M1[1][1]*M2[1][1]

    v1=M1[0][0]*M2[0][2]+M1[0][1]*M2[1][2]+M1[0][2]
    v2=M1[1][0]*M2[0][2]+M1[1][1]*M2[1][2]+M1[1][2]
    return [[a11,a12,v1],[a21,a22,v2]]

def writeTransformToNode(mat,node):
    node.set("transform", formatTransform(mat))

def applyTransformToNode(mat,node):
    m=parseTransform(node.get("transform"))
    writeTransformToNode(composeTransform(mat,m), node)

def applyTransformToPoint(mat,pt):
    x=mat[0][0]*pt[0]+mat[0][1]*pt[1]+mat[0][2]
    y=mat[1][0]*pt[0]+mat[1][1]*pt[1]+mat[1][2]
    pt[0]=x
    pt[1]=y

def applyTransformToPath(mat,path):
    for comp in path:
        for ctl in comp:
            for pt in ctl:
                applyTransformToPoint(mat,pt)

####################################################################
##-- Some functions to compute a rough bbox of a given list of objects.
##-- this should be shipped out in an separate file...

def boxunion(b1,b2):
    if b1 is None:
        return b2
    elif b2 is None:
        return b1    
    else:
        return((min(b1[0],b2[0]),max(b1[1],b2[1]),min(b1[2],b2[2]),max(b1[3],b2[3])))

def roughBBox(path):
    xmin,xMax,ymin,yMax=path[0][0][0][0],path[0][0][0][0],path[0][0][0][1],path[0][0][0][1]
    for pathcomp in path:
        for ctl in pathcomp:
           for pt in ctl:
               xmin=min(xmin,pt[0])
               xMax=max(xMax,pt[0])
               ymin=min(ymin,pt[1])
               yMax=max(yMax,pt[1])
    return xmin,xMax,ymin,yMax

def computeBBox(aList,mat=[[1,0,0],[0,1,0]]):
    bbox=None
    for node in aList:
        m = parseTransform(node.get('transform'))
        m = composeTransform(mat,m)
        #TODO: text not supported!

        if node.tag == inkex.addNS('rect','svg'):
                A=[0,0]
                B=[0,0]
                A[0] = float(node.get('x'))
                A[1] = float(node.get('y'))
                B[0] = A[0]+float(node.get('width'))
                B[1] = A[1]+float(node.get('height'))
                applyTransformToPoint(mat, A)
                applyTransformToPoint(mat, B)
                bbox = min(A[0], B[0]), max(A[0], B[0]), min(A[1], B[1]), max(A[1], B[1])
        if node.get("d"):
            d = node.get('d')
            p = cubicsuperpath.parsePath(d)
            applyTransformToPath(m,p)
            bbox=boxunion(roughBBox(p),bbox)

        if  node.tag == inkex.addNS('use','svg') or node.tag=='use':
            refid=node.get(inkex.addNS('href','xlink'))
            path = '//*[@id="%s"]' % refid[1:]
            refnode = node.getroottree().xpath(path, namespaces=inkex.NSS)
            bbox=boxunion(computeBBox(refnode,m),bbox)
            
        bbox=boxunion(computeBBox(node,m),bbox)
    return bbox
